### PR TITLE
feat: Integrate EDGE signaling

### PR DESCRIPTION
**Pull request by mykola-vrmchk**

### CODE REVIEW

File: packages/core/mesh/edge-client/src/client.ts [hunk 1]
```diff
@@ -18,17 +19,13 @@ const DEFAULT_TIMEOUT = 5_000;
...
-/**
- *
- */
-export interface EdgeConnection {
+export interface EdgeConnection extends Required<Lifecycle> {
   get info(): any;
   get identityKey(): PublicKey;
   get deviceKey(): PublicKey;
   get isOpen(): boolean;
+  setIdentity({ deviceKey, identityKey }: { deviceKey: PublicKey; identityKey: PublicKey }): void;
```
Comment (dmaretskyi) [inline packages/core/mesh/edge-client/src/client.ts:27]: are you planning to change those to strings?

File: packages/core/mesh/edge-client/src/client.ts [hunk 2]
```diff
@@ -42,16 +39,20 @@ export type MessengerConfig = {
  * Messenger client.
  */
 // TODO(dmaretskyi): Rename EdgeClient.
-export class EdgeClient implements EdgeConnection {
+// TODO(mykola): Handle reconnections.
+export class EdgeClient extends Resource implements EdgeConnection {
   private readonly _listeners: Set<MessageListener> = new Set();
```
Comment (dmaretskyi) [inline packages/core/mesh/edge-client/src/client.ts:44]:
```suggestion
  private readonly _listeners = new Set<MessageListener>();
```

File: packages/sdk/client-protocol/src/edge.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+/**
+ * Feature flags for edge services.
+ */
+export const EDGE_FEATURES = {
```
Comment (yarolegovich) [inline packages/sdk/client-protocol/src/edge.ts:8]: todo: move to config?

File: packages/core/mesh/edge-client/src/client.ts [hunk 4]
```diff
@@ -84,28 +96,29 @@ export class EdgeClient implements EdgeConnection {
   /**
    * Open connection to messaging service.
    */
-  public async open(): Promise<boolean> {
-    invariant(!this._ws);
+  protected override async _open() {
+    if (this._ws) {
```
Comment (dmaretskyi) [inline packages/core/mesh/edge-client/src/client.ts:100]: `_ws` is set to undefined on close. When would this check trigger?

File: packages/core/mesh/edge-client/src/client.ts [hunk 5]
```diff
@@ -76,6 +77,17 @@ export class EdgeClient implements EdgeConnection {
...
 
+  setIdentity({ deviceKey, identityKey }: { deviceKey: PublicKey; identityKey: PublicKey }) {
+    // TODO(mykola): Make synchronous. Manual close should cancel the operation.
+    this._deviceKey = deviceKey;
+    this._identityKey = identityKey;
+    this.close()
+      .then(async () => {
+        await this.open();
+      })
+      .catch(log.catch);
```
Comment (dmaretskyi) [inline packages/core/mesh/edge-client/src/client.ts:88]:
```suggestion
      .catch(err => log.catch(err));
```

passing log functions like this prevents the log metadata from being inserted (filenames, line numbers)

File: packages/core/mesh/edge-client/src/client.ts [hunk 6]
```diff
@@ -76,6 +77,17 @@ export class EdgeClient implements EdgeConnection {
     return !!this._ws;
   }
 
+  setIdentity({ deviceKey, identityKey }: { deviceKey: PublicKey; identityKey: PublicKey }) {
+    // TODO(mykola): Make synchronous. Manual close should cancel the operation.
```
Comment (dmaretskyi) [inline packages/core/mesh/edge-client/src/client.ts:81]: still valid?

File: packages/core/mesh/edge-client/src/client.ts [hunk 7]
```diff
@@ -76,6 +77,17 @@ export class EdgeClient implements EdgeConnection {
     return !!this._ws;
```
Comment (dmaretskyi) [inline packages/core/mesh/edge-client/src/client.ts:77]: maybe use Resource _lifecycleState instead?

File: packages/core/mesh/network-manager/src/network-manager.ts [hunk 8]
```diff
@@ -56,7 +56,7 @@ export type SwarmOptions = {
 export type SwarmNetworkManagerOptions = {
   transportFactory: TransportFactory;
   signalManager: SignalManager;
-  log?: boolean; // Log to devtools.
+  shouldLog?: boolean; // Log to devtools.
```
Comment (dmaretskyi) [inline packages/core/mesh/network-manager/src/network-manager.ts:59]:
```suggestion
  enableDevtoolsLogging?: boolean; // Log to devtools.
```

File: packages/core/mesh/network-manager/src/swarm/connection.ts [hunk 9]
```diff
@@ -125,8 +126,8 @@ export class Connection {
 
   constructor(
     public readonly topic: PublicKey,
-    public readonly ownId: PublicKey, // TODO(burdon): peerID?
-    public readonly remoteId: PublicKey,
+    public readonly own: PeerInfo,
```
Comment (dmaretskyi) [inline packages/core/mesh/network-manager/src/swarm/connection.ts:129]:
```suggestion
    public readonly ownInfo: PeerInfo,
```

File: packages/core/mesh/network-manager/src/swarm/connection.ts [hunk 10]
```diff
@@ -125,8 +126,8 @@ export class Connection {
...
+    public readonly remote: PeerInfo,
```
Comment (dmaretskyi) [inline packages/core/mesh/network-manager/src/swarm/connection.ts:130]:
```suggestion
    public readonly remoteInfo: PeerInfo,
```

File: packages/core/mesh/network-manager/src/swarm/connection.ts (hunk 9 above)
Comment (dmaretskyi) [reply packages/core/mesh/network-manager/src/swarm/connection.ts:129]: Prefer not to use adjectives as identifiers like that

Comment (dmaretskyi) [reply packages/core/mesh/network-manager/src/swarm/connection.ts:129]: `localInfo` would be an even better name

**Review by dmaretskyi — APPROVED** [review]

File: packages/core/mesh/messaging/src/signal-manager/edge-signal-manager.ts [hunk 11]
```diff
@@ -14,48 +16,61 @@ import {
...
 import { ComplexMap, ComplexSet } from '@dxos/util';
 
-import { type Message, type SignalMethods } from '../signal-methods';
+import { type SignalManager } from './signal-manager';
+import { type PeerInfo, type Message, type SwarmEvent, PeerInfoHash } from '../signal-methods';
 
 const SWARM_SERVICE_ID = 'swarm';
-export class EdgeSignal extends Resource implements SignalMethods {
-  public swarmEvent = new Event<{ topic: PublicKey; swarmEvent: SwarmEvent }>();
+const SIGNAL_SERVICE_ID = 'signal';
```
Comment (dmaretskyi) [inline packages/core/mesh/messaging/src/signal-manager/edge-signal-manager.ts:25]: There shouldn't be a singal service.
`serviceId` is used when you want to route a message to a service inside EDGE (e.g. replicator or swarm).
For messages between peers we should leave this field empty.

File: packages/core/mesh/messaging/src/signal-manager/edge-signal-manager.ts [hunk 12]
```diff
@@ -14,48 +16,61 @@ import {
...
   type Message as EdgeMessage,
 } from '@dxos/protocols/buf/dxos/edge/messenger_pb';
-import { type SwarmEvent } from '@dxos/protocols/proto/dxos/mesh/signal';
 import { ComplexMap, ComplexSet } from '@dxos/util';
 
-import { type Message, type SignalMethods } from '../signal-methods';
+import { type SignalManager } from './signal-manager';
+import { type PeerInfo, type Message, type SwarmEvent, PeerInfoHash } from '../signal-methods';
 
 const SWARM_SERVICE_ID = 'swarm';
```
[the remark this answers is not in the document]
Comment (dmaretskyi) [reply packages/core/mesh/messaging/src/signal-manager/edge-signal-manager.ts:24]: 👍

File: packages/core/mesh/edge-client/src/client.ts (hunk 1 above)
Comment (mykola-vrmchk) [reply packages/core/mesh/edge-client/src/client.ts:27]: Yes

File: packages/core/mesh/messaging/src/messenger.ts [hunk 13]
```diff
@@ -36,12 +36,12 @@ export class Messenger {
...
   // { peerId, payloadType } => listeners set
-  private readonly _listeners = new ComplexMap<{ peerId: PublicKey; payloadType: string }, Set<OnMessage>>(
-    ({ peerId, payloadType }) => peerId.toHex() + payloadType,
+  private readonly _listeners = new ComplexMap<{ peerId: string; payloadType: string }, Set<OnMessage>>(
+    ({ peerId, payloadType }) => peerId + payloadType,
   );
 
   // peerId => listeners set
-  private readonly _defaultListeners = new ComplexMap<PublicKey, Set<OnMessage>>(PublicKey.hash);
+  private readonly _defaultListeners = new Map<string, Set<OnMessage>>();
```
[the remark this answers is not in the document]
Comment (mykola-vrmchk) [reply packages/core/mesh/messaging/src/messenger.ts:44]: It is safe assuming that all deviceKeys are unique

**Pull request merged**